static/app/components/forms/compactSelect.tsx

```diff
@@ -0,0 +1,312 @@
+  trigger,
+  triggerLabel,
+  triggerProps,
+  className,
+  renderWrapAs,
+  // Overlay props
+  offset = 8,
+  crossOffset = 0,
+  containerPadding = 8,
+  placement = 'bottom left',
+  closeOnSelect = true,
+  shouldCloseOnBlur = true,
+  isDismissable = true,
+  menuTitle,
+  ...props
+}: Props) {
+  // Manage the dropdown menu's open state
+  const isDisabled = disabledProp || options?.length === 0;
+  const triggerRef = useRef<HTMLButtonElement>(null);
+  const state = useMenuTriggerState(props);
+  const {menuTriggerProps} = useMenuTrigger(
+    {type: 'listbox', isDisabled},
+    state,
+    triggerRef
+  );
+  const {buttonProps} = useButton(
+    {onPress: () => state.toggle(), isDisabled, ...menuTriggerProps},
+    triggerRef
+  );
+
+  // Control the overlay's position
+  const overlayRef = useRef(null);
+  const {overlayProps} = useOverlay(
+    {
+      onClose: state.close,
+      isOpen: state.isOpen,
+      shouldCloseOnBlur,
+      isDismissable,
+    },
+    overlayRef
+  );
+  const {overlayProps: positionProps} = useOverlayPosition({
+    targetRef: triggerRef,
+    overlayRef,
+    offset,
+    crossOffset,
+    placement,
+    containerPadding,
+    isOpen: state.isOpen,
+  });
+
+  // Keep an internal copy of the current select value and update the control
+  // button's label when the value changes
+  const [internalValue, setInternalValue] = useState(valueProp ?? defaultValue);
+  const [label, setLabel] = useState(getLabel(valueProp ?? null));
+
+  // Update the button label when the value changes
+  function getLabel(newValue): React.ReactNode {
+    const valueSet = Array.isArray(newValue) ? newValue : [newValue];
+    const optionSet = valueSet.map(val => options.find(opt => opt.value === val));
+    const firstOptionLabel = optionSet[0] ? optionSet[0]?.label : '';
+
+    return (
+      <Fragment>
+        <ButtonLabel>{firstOptionLabel}</ButtonLabel>
+        {optionSet.length > 1 && <StyledBadge text={`+${optionSet.length - 1}`} />}
+      </Fragment>
+    );
+  }
+
+  useEffect(() => {
+    const newValue = valueProp ?? internalValue;
+    const newLabel = getLabel(newValue);
+    setLabel(newLabel);
+  }, [valueProp ?? internalValue]);
+
+  // Calculate & update the trigger button's width, to be used as the
+  // overlay's min-width
+  const [triggerWidth, setTriggerWidth] = useState<number>();
+  useEffect(() => {
+    // Wait until the trigger label has been updated before calculating the
+    // new width
+    setTimeout(() => {
+      const newTriggerWidth = triggerRef.current?.offsetWidth;
+      newTriggerWidth ?? setTriggerWidth(newTriggerWidth);
+    }, 0);
+  }, [triggerRef.current, internalValue]);
+
+  function onValueChange(option) {
+    const newValue = Array.isArray(option) ? option.map(opt => opt.value) : option?.value;
+    setInternalValue(newValue);
+    onChange?.(option);
+
+    if (closeOnSelect && !multiple) {
+      state.close();
+    }
+  }
+
+  function renderTrigger() {
```

Comment:
@vuluongj20 How do you feel about moving these to separate components or inlining them into the return call? This has a small benefit of making it easier to just read the render call and not have to jump back and forth between function return and renderTrigger call?

Reply:
Yeah, I think using `!trigger ? <DropdownButton/> : null` is a super common react pattern (more common that having render* functions) so I would just go with that